### PR TITLE
Add validation for locked feature gate

### LINKED ISSUE #5362: Locked feature gates are not validated
**How to categorize this issue?**

/area ops-productivity
/kind bug

**What happened**:
We have seen cases where end-users were disabling feature gates and later upgrading to a Kubernetes version which does no longer allow disabling such feature gates (because they were promoted to GA and are 'locked' now).

This resulted in a crash-looping `kube-apiserver`:

```
Flow "Shoot cluster deletion" encountered task errors: [task "Waiting until Kubernetes API server reports readiness" failed: retry failed with context deadline exceeded, last error: kube-apiserver deployment has outdated replicas, logs of newest pod:
Error: invalid argument "BoundServiceAccountTokenVolume=false,CSIMigration=true,CSIMigrationAWS=true,InTreePluginAWSUnregister=true" for "--feature-gates" flag: cannot set feature gate BoundServiceAccountTokenVolume to false, feature is locked to true
]
```

If end-users now delete the `Shoot` then they cannot even help themselves (via removing/changing the feature gate configuration) since any spec changes are forbidden if a deletion timestamp is present.

**What you expected to happen**:
Gardener to forbid upgrading clusters if the feature gate configuration is invalid.

**How to reproduce it (as minimally and precisely as possible)**:
1. Create a `Shoot` with Kubernetes 1.21 and set `BoundServiceAccountTokenVolume=false`.
2. Upgrade the `Shoot` to Kubernetes 1.22 (in which `BoundServiceAccountTokenVolume` was promoted to GA and is locked to `True`)
3. See the `kube-apiserver` failing.

**Anything else we need to know?**:
We already validate feature gates w.r.t. "added in version" and "removed in version" (https://github.com/gardener/gardener/blob/master/pkg/utils/kubernetes/featuregates.go). We should extend this by also allowing describing "locked in version" and validate that those constraints are met.

**Environment**:

- Gardener version: v1.38.4

### PR DESCRIPTION
**Pull request by acumino**
**How to categorize this PR?**

/area ops-productivity
/kind bug

**What this PR does / why we need it**:
This PR adds validation for the locked feature gates. Now Gardener forbids upgrading clusters if the feature gate is locked to default and the user wants to set a value that does not default.

**Which issue(s) this PR fixes**:
Fixes #5362

**Special notes for your reviewer**:
This PR adds doesn't add validation for the feature gate which has been removed in any of the Kubernetes versions supported by the Gardener.

**Release note**:

```other operator
NONE
```

### CODE REVIEW

File: pkg/utils/validation/features/featuregates_test.go [hunk 1]
```diff
@@ -108,6 +108,11 @@ var _ = Describe("featuregates", func() {
 				"BadValue": Equal("Foo"),
 				"Detail":   Equal("unknown feature gate Foo"),
 			})))),
+			Entry("setting non-default value for lockedfeaturegate", map[string]bool{"EndpointSlice": false}, "1.21.5", ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{
```
Comment (rfranzke) [inline pkg/utils/validation/features/featuregates_test.go]:
```suggestion
			Entry("setting non-default value for locked feature gate", map[string]bool{"EndpointSlice": false}, "1.21.5", ConsistOf(PointTo(MatchFields(IgnoreExtras, Fields{
```

**Review by rfranzke — CHANGES_REQUESTED** [review]
Thanks!

Can you please adapt the script to fetch this information (https://github.com/gardener/gardener/blob/master/hack/compare-k8s-feature-gates.sh), the explanation doc string (https://github.com/gardener/gardener/blob/96dd4b835cdd80803560060d49847249ddcc68f7/pkg/utils/validation/features/featuregates.go#L27-L31) and the docs for supporting a new Kubernetes version (https://github.com/gardener/gardener/blob/master/docs/development/new-kubernetes-version.md#adapting-gardener)?

**Comment by rfranzke** [comment]
The explanation doc string is not yet adapted.

**Comment by acumino** [comment]
> The explanation doc string is not yet adapted.

The function I have added is not exported do we still need Doc string?

**Comment by rfranzke** [comment]
I am talking about:

> the explanation doc string (https://github.com/gardener/gardener/blob/96dd4b835cdd80803560060d49847249ddcc68f7/pkg/utils/validation/features/featuregates.go#L27-L31)

File: docs/development/new-kubernetes-version.md [hunk 2]
```diff
@@ -56,9 +56,10 @@ Run the [`.ci/check-and-release`](https://github.com/gardener/hyperkube/blob/mas
 - Maintain the Kubernetes feature gates used for validation of `Shoot` resources:
   - The feature gates are maintained in [this](https://github.com/gardener/gardener/blob/master/pkg/utils/validation/features/featuregates.go) file.
   - To maintain this list for new Kubernetes versions, run `hack/compare-k8s-feature-gates.sh <old-version> <new-version>` (e.g. `hack/compare-k8s-feature-gates.sh v1.22 v1.23`).
-  - It will present 2 lists of feature gates: those added and those removed in `<new-version>` compared to `<old-version>`.
+  - It will present 3 lists of feature gates: those added and those removed in `<new-version>` compared to `<old-version>` and feature gates that got locked to default in `<new-version>`.
   - Add all added feature gates to the map with `<new-version>` as `AddedInVersion` and no `RemovedInVersion`.
   - For any removed feature gates, add `<new-version>` as RemovedInVersion to the already existing feature gate in the map.
+  - For feature gates locked to default, add `<new-version>` as LockedToDefaultInVersion to the already existing feature gate in the map.
```
Comment (rfranzke) [inline docs/development/new-kubernetes-version.md]:
```suggestion
  - For feature gates locked to default, add `<new-version>` as `LockedToDefaultInVersion` to the already existing feature gate in the map.
```

**Review by rfranzke — COMMENTED** [review]
Nit

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**